test/integration/websocket_integration_test.cc

```diff
@@ -21,7 +21,76 @@ using testing::MatchesRegex;
+                                 {"upgrade", upgrade_type},
+                                 {"content-length", "0"}};
+}
+
+void validateUpgradeRequestHeaders(const Http::HeaderMap& original_proxied_request_headers,
+                                   const Http::HeaderMap& original_request_headers,
+                                   bool old_style) {
+  Http::TestHeaderMapImpl proxied_request_headers(original_proxied_request_headers);
+  if (proxied_request_headers.ForwardedProto()) {
+    ASSERT_STREQ(proxied_request_headers.ForwardedProto()->value().c_str(), "http");
+    proxied_request_headers.removeForwardedProto();
+  }
+  if (proxied_request_headers.Scheme()) {
+    ASSERT_STREQ(proxied_request_headers.Scheme()->value().c_str(), "http");
+    proxied_request_headers.removeScheme();
+  }
+
+  if (!old_style) {
+    // Check for and remove headers added by default for HTTP requests.
+    ASSERT_TRUE(proxied_request_headers.RequestId() != nullptr);
+    ASSERT_TRUE(proxied_request_headers.EnvoyExpectedRequestTimeoutMs() != nullptr);
+    proxied_request_headers.removeEnvoyExpectedRequestTimeoutMs();
+  } else {
+    // Check for and undo the path rewrite.
+    ASSERT_STREQ(proxied_request_headers.Path()->value().c_str(), "/websocket");
+    proxied_request_headers.Path()->value().append("/test", 5);
+    ASSERT_STREQ(proxied_request_headers.EnvoyOriginalPath()->value().c_str(), "/websocket/test");
+    proxied_request_headers.removeEnvoyOriginalPath();
+  }
+  // If no content length is specified, the HTTP codec should add a chunked encoding header.
+  if (original_request_headers.ContentLength() == nullptr) {
+    ASSERT_STREQ(proxied_request_headers.TransferEncoding()->value().c_str(), "chunked");
+    proxied_request_headers.removeTransferEncoding();
+  }
+  proxied_request_headers.removeRequestId();
+
+  EXPECT_EQ(proxied_request_headers, original_request_headers);
+}
+
+void validateUpgradeResponseHeaders(const Http::HeaderMap& original_proxied_response_headers,
+                                    const Http::HeaderMap& original_response_headers,
+                                    bool old_style) {
+  Http::TestHeaderMapImpl proxied_response_headers(original_proxied_response_headers);
+  if (!old_style) {
+    // Check for and remove headers added by default for HTTP responses.
+    ASSERT_TRUE(proxied_response_headers.Date() != nullptr);
+    ASSERT_TRUE(proxied_response_headers.Server() != nullptr);
+    ASSERT_STREQ(proxied_response_headers.Server()->value().c_str(), "envoy");
+    proxied_response_headers.removeDate();
+    proxied_response_headers.removeServer();
+  }
+  // If no content length is specified, the HTTP codec should add a chunked encoding header.
+  if (original_response_headers.ContentLength() == nullptr) {
+    ASSERT_STREQ(proxied_response_headers.TransferEncoding()->value().c_str(), "chunked");
+    proxied_response_headers.removeTransferEncoding();
+  }
```

Comment:
Looks like the same check on ContentLength is repeated for request and response. Worth factoring it out?